llumnix/global_scheduler/migration_scheduler.py

```diff
@@ -78,74 +79,106 @@ def remove_instance(self, instance_id: str) -> None:
         self.instance_id_set.remove(instance_id)
         self.num_instances = len(self.instance_id_set)
 
-    def _sort_instance_infos(self,
-                             descending: bool = True) -> None:
-        instance_infos: List[InstanceInfo] = list(self.instance_info.values())
-        key_attr = 'instance_load_migrate'
-        self.sorted_instance_infos = sorted(
-            instance_infos,
-            key=lambda instance_info: getattr(instance_info, key_attr),
-            reverse=descending
-        )
-
-class FilteringInstanceInfosPolicy(ABC):
-    def __init__(self,
-                 migrate_out_load_threshold: float) -> None:
-        self.migrate_out_load_threshold = migrate_out_load_threshold
-        self.filter_instances_rules = {
-            PairMigrationConstraints.NO_CONSTRAINTS: (InstanceType.NO_CONSTRAINTS, InstanceType.NO_CONSTRAINTS),
-            PairMigrationConstraints.DECODING_2_DECODING: (InstanceType.DECODE, InstanceType.DECODE),
-            PairMigrationConstraints.PREFILL_2_DECODING: (InstanceType.PREFILL, InstanceType.DECODE),
-        }
-
-    def filter_instances(self, sorted_instance_infos: List[InstanceInfo],
-                         pair_migration_type: PairMigrationConstraints = None) -> Dict[str, InstanceInfo]:
-        src_type, dst_type = self.filter_instances_rules[pair_migration_type]
-        filtered_src_instance_infos = [info for info in sorted_instance_infos if info.instance_type == src_type]
-        filtered_dst_instance_infos = [info for info in sorted_instance_infos if info.instance_type == dst_type]
-        src_instance_infos = self.filter_src_instances(filtered_src_instance_infos)
-        dst_instance_infos = self.filter_dst_instances(filtered_dst_instance_infos)
-        return src_instance_infos, dst_instance_infos
-
+MIGRATION_FILTER_POLICY_MAPPING = {
+    PairMigrationConstraints.NO_CONSTRAINTS: 'constraint',
+    PairMigrationConstraints.DECODING_2_DECODING: 'constraint',
+    PairMigrationConstraints.PREFILL_2_DECODING: 'relax',
+}
+
+class MigrationFilter(ABC):
+    def __init__(self, filter_config: MigrationFilterConfig) -> None:
+        self.filter_config = filter_config
+        self.addition_src_filter: Callable[[InstanceInfo], bool] = None
+        self.addition_dst_filter: Callable[[InstanceInfo], bool] = None
+        self.default_filter = MigrationFilterPolicyFactory.get_policy('instancetype')
```

Comment:
In original design, selecting instances involves two steps: 1. Filter out prefill/decoding instances (InstanceTypeFilter). 2. Filter based on instanceInfo (num_killed_requests/load).
Steps 1 and 2 are split into two independent strategies in this PR, meaning only one or the other can be chosen? I am concerned about its reasonability, since two strategies have an incremental logic and do not constrain each other. By using a hierarchical filtering approach, it can eliminate unreasonable migration pairs. It seems that only choosing `InstanceTypeFilter`, many unnecessary coroutines will be started.

Reply:
see filter_instances.src_filter_conditions and dst_filter_conditions.

multi-filter can be set.